jack-test/test/java/com/rapleaf/jack/transaction/TestDbTransactorImpl.java

```diff
@@ -38,6 +44,43 @@ public void cleanup() throws Exception {
     executorService = null;
   }
 
+  @Test
+  public void testConnectionsReturnedOnRollbackFailure() {
+    MysqlDatabaseConnection realConnection = new MysqlDatabaseConnection("database1");
+    BaseDatabaseConnection mBaseDatabaseConnection = Mockito.spy(realConnection);
+    IDatabase1 database1 =
+        new Database1Impl(
+            mBaseDatabaseConnection,
+            new DatabasesImpl(mBaseDatabaseConnection),
+            new NoOpAction()
+        );
+
+    Mockito.doAnswer(invocationOnMock1 -> {
+      throw new SQLRecoverableException("Commit failure");
+    }).when(mBaseDatabaseConnection).commit();
+
+    Mockito.doAnswer(invocationOnMock -> {
+      throw new SQLRecoverableException("Rollback failure");
+    }).when(mBaseDatabaseConnection).rollback();
+
+    DbPoolManager<IDatabase1> dbPoolManager = Mockito.spy(new DbPoolManager<>(() -> database1, 1,
+        1, 100, Integer.MAX_VALUE, false
+    ));
+    TransactorImpl<IDatabase1> transactor = new TransactorImpl<>(dbPoolManager, false);
+
+    try {
+      transactor.queryAsTransaction(db -> db.users().find(1));
+      fail();
+    } catch (Exception e) {
+      // This should fail
```

Comment:
Probably need to fix up the wording here. It currently seems like its implying something should be failing inside this catch block instead of implying that this is the correct behavior.

Reply:
Changed it. Hopefully it's more clear now.